src/main/scala/org/constellation/domain/healthcheck/HealthCheckConsensus.scala

```diff
@@ -0,0 +1,569 @@
+  checkedId: Id,
+  ownId: Id,
+  roundId: RoundId,
+  initialRoundPeers: Map[Id, PeerData],
+  healthCheckStatus: PeerHealthCheckStatus,
+  startedAtSecond: Long,
+  val isKeepUpRound: Boolean
+)(
+  apiClient: ClientInterpreter[F],
+  unboundedBlocker: Blocker
+)(implicit F: Concurrent[F], CS: ContextShift[F], C: Clock[F]) {
+  import HealthCheckConsensus._
+
+  implicit val roundIdOrdering: Ordering[RoundId] = Ordering.by[RoundId, String](_.id)
+  implicit val roundIdOrder: Order[RoundId] = Order.fromOrdering
+  implicit val peerDataOrdering: Ordering[PeerData] = Ordering.by[PeerData, String](_.peerMetadata.id.hex)
+  implicit val peerDataOrder: Order[PeerData] = Order.fromOrdering
+
+  val logger: SelfAwareStructuredLogger[F] = Slf4jLogger.getLogger[F]
+  val fetchProposalAfter = 120L
+  val removeUnresponsivePeersWithParallelRoundAfter = 60L
+
+  private[healthcheck] val roundIds: Ref[F, NonEmptySet[RoundId]] = Ref.unsafe(NonEmptySet.one(roundId))
+  private[healthcheck] val roundPeers: Ref[F, Map[Id, RoundData]] = {
+    Ref.unsafe(
+      initialRoundPeers
+        .mapValues(pd => RoundData(pd.asRight[ProxyPeers]))
+    )
+  }
+  private[healthcheck] val leftDuringConsensus: Ref[F, Set[Id]] = Ref.unsafe(Set.empty)
+  private[healthcheck] val removedUnresponsivePeersWithParallelRound: Ref[F, Set[Id]] = Ref.unsafe(Set.empty)
+  private[healthcheck] val ownConsensusHealthStatus: Ref[F, ConsensusHealthStatus] = Ref.unsafe(
+    ConsensusHealthStatus(
+      checkedId,
+      ownId,
+      roundId,
+      healthCheckStatus,
+      initialRoundPeers.mapValues(_.peerMetadata.nodeState)
+    )
+  )
+  private[healthcheck] val parallelRounds: Ref[F, Map[Id, Set[RoundId]]] = Ref.unsafe(Map.empty)
+
+  private val sendingProposalLock = createSemaphore()
+  private val fetchingProposalLock = createSemaphore()
+
+  def getTimeInSeconds(): F[Long] = C.monotonic(SECONDS)
+
+  def start(): F[Fiber[F, Unit]] = F.start(sendProposal())
+
+  def getOwnPerceivedHealthStatus(): F[ConsensusHealthStatus] = ownConsensusHealthStatus.get
+
+  def getRemovedUnresponsivePeersWithParallelRound(): F[Set[Id]] = removedUnresponsivePeersWithParallelRound.get
+
+  def addRoundIds(ids: Set[RoundId]): F[Unit] =
+    roundIds.modify(rounds => (NonEmptySet(rounds.head, rounds.tail ++ ids), ()))
+
+  def getRoundIds(): F[NonEmptySet[RoundId]] = roundIds.get
+
+  def getParallelRounds(): F[Map[Id, Set[RoundId]]] = parallelRounds.get
+
+  def addParallelRounds(id: Id, roundIds: Set[RoundId]): F[Unit] = parallelRounds.modify { rounds =>
+    val updatedIdRounds = rounds.get(id).map(r => r ++ roundIds).getOrElse(roundIds)
+
+    (rounds + (id -> updatedIdRounds), ())
+  }
+
+  def getRoundPeers(): F[Map[Id, RoundData]] = roundPeers.get
+
+  def addRoundPeer(id: Id, peerData: PeerData): F[Unit] =
+    roundPeers.modify(peers => (peers + (id -> RoundData(peerData.asRight[ProxyPeers])), ()))
+
+  def removePeer(id: Id): F[Unit] = roundPeers.modify(peers => (peers - id, ()))
```

Comment:
IMO the name of this fn is misleading. When I was reading another part of code I originally though it should remove the peer but looks like it removes it only from this particular round. Maybe `ignorePeerInRound` or sth with `ignore` not `remove` ?

Reply:
It's in the consensus round logic so it is regarding this single consensus round. Possibly it could be named better or done as a part of other functions which it works in conjunction with. When the node leaves we put it in a map of leftDuringConsensus, and then removing from roundPeers with this function.